File: explore/queue_and_stack/queue/my-circular-queue.c
#include <stdlib.h>
#include <stdbool.h>

typedef struct MyCircularQueue
{
    int *data;
    int head;
    int tail;
    int size;
} MyCircularQueue;

/** Initialize your data structure here. Set the size of the queue to be k. */
MyCircularQueue *myCircularQueueCreate(int k)
{
    MyCircularQueue *obj;
    if ((obj = malloc(sizeof(*obj))) == NULL)
        return NULL;
    if ((obj->data = malloc(sizeof(int) * k)) == NULL)
    {
        free(obj);
        return NULL;
    }
    obj->head = obj->tail = -1;
    obj->size = k;
    return obj;
}

/** Checks whether the circular queue is empty or not. */
bool myCircularQueueIsEmpty(MyCircularQueue *obj)
{
    return obj->head == -1;
}

/** Checks whether the circular queue is full or not. */
bool myCircularQueueIsFull(MyCircularQueue *obj)
{
    return (obj->tail + 1) % obj->size == obj->head;
}

/** Insert an element into the circular queue. Return true if the operation is successful. */
bool myCircularQueueEnQueue(MyCircularQueue *obj, int value)
{
    if (myCircularQueueIsFull(obj))
        return false;
    if (myCircularQueueIsEmpty(obj))
        obj->head = 0;
    obj->tail = (obj->tail + 1) % obj->size;
    obj->data[obj->tail] = value;
    return true;
}

/** Delete an element from the circular queue. Return true if the operation is successful. */
bool myCircularQueueDeQueue(MyCircularQueue *obj)
{
	if (myCircularQueueIsEmpty(obj))
		return false;
	if (obj->head == obj->tail)
		obj->head = obj->tail = -1;
	else 
		obj->head = (obj->head + 1) % obj->size;
	return true;
}

/** Get the front item from the queue. */
int myCircularQueueFront(MyCircularQueue *obj)
{
    if (myCircularQueueIsEmpty(obj))
        return -1;
    return obj->data[obj->head];
}

/** Get the last item from the queue. */
int myCircularQueueRear(MyCircularQueue *obj)
{
    if (myCircularQueueIsEmpty(obj))
        return -1;
    return obj->data[obj->tail];
}

void myCircularQueueFree(MyCircularQueue *obj)
{
    free(obj->data);
    free(obj);
}

File: explore/queue_and_stack/queue/MyCircularQueue.java
class MyCircularQueue {

    private int[] data;
    private int head;
    private int tail;
    private int size;
    private int length;

    /** Initialize your data structure here. Set the size of the queue to be k. */
    public MyCircularQueue(int k) {
        size = k;
        length = 0;
        data = new int[size];
        head = tail = -1;
    }
    
    /** Insert an element into the circular queue. Return true if the operation is successful. */
    public boolean enQueue(int value) {
        if (length == size) {
            return false;
        }
        if (head == -1) {
            data[0] = value;
            tail = head = 0;
        } else {
            tail = (tail + 1) % size;
            data[tail] = value;
        }
        length++;
        return true;
    }
    
    /** Delete an element from the circular queue. Return true if the operation is successful. */
    public boolean deQueue() {
        if (length == 0) {
            return false;
        }
        head = head + 1 % size;
        length--;
        if (length == 0) {
            head = tail = -1;
        }
        return true;
    }
    
    /** Get the front item from the queue. */
    public int Front() {
        if (length == 0) {
            return -1;
        }
        return data[head];
    }
    
    /** Get the last item from the queue. */
    public int Rear() {
        if (length == 0) {
            return -1;
        }
        return data[tail];
    }
    
    /** Checks whether the circular queue is empty or not. */
    public boolean isEmpty() {
        return length == 0;
    }
    
    /** Checks whether the circular queue is full or not. */
    public boolean isFull() {
        return length == size;
    }
}

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue obj = new MyCircularQueue(k);
 * boolean param_1 = obj.enQueue(value);
 * boolean param_2 = obj.deQueue();
 * int param_3 = obj.Front();
 * int param_4 = obj.Rear();
 * boolean param_5 = obj.isEmpty();
 * boolean param_6 = obj.isFull();
 */